torch/csrc/jit/codegen/cuda/kernel.cpp

```diff
@@ -25,58 +26,135 @@ static int ceilDiv(const int a, const int b) {
   return (a + b - 1) / b;
 }
 
+// IO data structure for kernel code;
+static auto typeinfo = R"(
+template<typename T, int N>
+struct Tensor {
+  T& operator[](int ind) {
+    return data[ind];
+  };
+
+  T* data;
+  int size[N];
+  int stride[N];
+};
+)";
+
+// include IO data structure for host code
+struct KernelArgumentHolder {
+  std::vector<void*> arguments;
+  std::vector<char> buffer;
+  void* buffer_ptr;
+
+  KernelArgumentHolder(size_t n_inputs, size_t n_dimension_per_tensor) {
+    arguments.reserve(n_inputs);
+    // We are being generous here on the allocated buffer;
+    buffer.resize(
+        n_inputs * (sizeof(void*) + 2 * sizeof(int) * n_dimension_per_tensor));
+    buffer_ptr = buffer.data();
+  }
+
+  void** args() {
+    return arguments.data();
+  }
+
+  // this should comply to the storage of Tensor object defined in typeinfo;
+  void push_tensor(
+      const at::Tensor& val,
+      c10::optional<at::IntArrayRef> broadcasted_size = c10::nullopt) {
+    arguments.push_back(buffer_ptr);
+
+    // passing address, type doesn't really matter here;
+    auto data_ptr = static_cast<void**>(buffer_ptr);
+    *data_ptr = val.data_ptr();
+    buffer_ptr += sizeof(char*);
+
+    if (broadcasted_size) {
+      auto b_dim = broadcasted_size->size();
+      auto o_dim = val.dim();
+      TORCH_CHECK(b_dim >= o_dim);
+      int* sizes = reinterpret_cast<int*>(buffer_ptr);
```

Comment:
Instead of doing all this manipulation, does it make sense to instantiate the template for dims 1 through max_dim, and select which one to use at runtime? Are you worried about the flexibility of this approach in the future?

Reply:
> instantiate the template for dims 1 through max_dim, and select which one to use at runtime

That was also my initial idea, just to keep the code simple
But I can't figure out how to cleanly do  `select which one to use at runtime`. The parts I don't like (don't know how to do cleanly): 1. a gigantic switch or MACRO to pick the instantiation with the right rank; 2. have a container class to host instances with different ranks;

In the end  I decided to just follow the legacy fuser (hence this implementation). This is generally safer and more flexible as it doesn't have that restriction of static `max_dim`.

Reply:
We can put this off, since we're doing so much RTTI I generally think this won't be too hard if we want to do it. We can keep it in mind for now.

The dispatch on RTTI is interesting, but once you get used to it, I think it's actually nice.